lib/srv/db/sqlserver/connect_test.go

```diff
@@ -116,3 +125,255 @@ func TestConnectorSelection(t *testing.T) {
+	err := os.WriteFile(krbConfPath, []byte(krb5Conf), 0664)
+	require.NoError(t, err)
+
+	for _, tt := range []struct {
+		desc         string
+		databaseSpec types.DatabaseSpecV3
+		errAssertion require.ErrorAssertionFunc
+	}{
+		{
+			desc: "AD-x509-Loads_and_fails_with_expired_cache",
+			databaseSpec: types.DatabaseSpecV3{
+				Protocol: defaults.ProtocolSQLServer,
+				URI:      "sqlserver:1443",
+				AD: types.AD{
+					LDAPCert:    mockCA,
+					KDCHostName: "kdc.example.com",
+					Krb5File:    krbConfPath,
+				},
+			},
+			// When using a non-Azure database, the connector should attempt to get a kinit client
+			errAssertion: func(t require.TestingT, err error, _ ...interface{}) {
+				require.Error(t, err)
+				// we can't get a new TGT without an actual kerberos implementation, so we are relying on the existing
+				// credentials cache being expired
+				require.ErrorContains(t, err, "cannot login, no user credentials available and no valid existing session")
+			},
+		},
+		{
+			desc: "AD-x509-Fails_to_load_with_bad_config",
+			databaseSpec: types.DatabaseSpecV3{
+				Protocol: defaults.ProtocolSQLServer,
+				URI:      "sqlserver:1443",
+				AD:       types.AD{},
+			},
+			// When using a non-Azure database, the connector should attempt to get a kinit client
+			errAssertion: func(t require.TestingT, err error, _ ...interface{}) {
+				require.Error(t, err)
+				// we can't get a new TGT without an actual kerberos implementation, so we are relying on the existing
+				// credentials cache being expired
+				require.ErrorIs(t, err, errBadKerberosConfig)
+			},
+		},
+		{
+			desc: "AD-x509-Fails_with_invalid_certificate",
+			databaseSpec: types.DatabaseSpecV3{
+				Protocol: defaults.ProtocolSQLServer,
+				URI:      "sqlserver:1443",
+				AD: types.AD{
+					LDAPCert:    "BEGIN CERTIFICATE",
+					KDCHostName: "kdc.example.com",
+					Krb5File:    krbConfPath,
+				},
+			},
+			// When using a non-Azure database, the connector should attempt to get a kinit client
+			errAssertion: func(t require.TestingT, err error, _ ...interface{}) {
+				require.Error(t, err)
+				// we can't get a new TGT without an actual kerberos implementation, so we are relying on the existing
+				// credentials cache being expired
+				require.ErrorIs(t, err, errBadCertificate)
+			},
+		},
+	} {
+		t.Run(tt.desc, func(t *testing.T) {
+			database, err := types.NewDatabaseV3(types.Metadata{
+				Name: uuid.NewString(),
+			}, tt.databaseSpec)
+			require.NoError(t, err)
+
+			databaseUser := "alice"
+			databaseName := database.GetName()
+			connector.DataDir = dir
+
+			connectorCtx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			resChan := make(chan error, 1)
+			go func() {
+				_, _, err = connector.Connect(connectorCtx, &common.Session{Database: database, DatabaseUser: databaseUser, DatabaseName: databaseName}, &protocol.Login7Packet{})
+				resChan <- err
+			}()
+
+			// Cancel the context to avoid dialing databases.
+			cancel()
```

Comment:
I think cancelling this here is racey/unpredictable. There's no guarantee that the go routine will be scheduled and start executing before `cancel()` is called. If the intention is to pass a cancelled context in, we should cancel it before kicking off the goroutine.

Reply:
Good point